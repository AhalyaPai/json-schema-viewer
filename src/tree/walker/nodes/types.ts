import { RegularNode } from './RegularNode';
import { ReferenceNode } from './ReferenceNode';
import { ErrorNode } from './ErrorNode';

export type SchemaNode = RegularNode | ReferenceNode | ErrorNode;

export enum SchemaNodeKind {
  Any = 'any',
  String = 'string',
  Number = 'number',
  Integer = 'integer',
  Boolean = 'boolean',
  Null = 'null',
  Array = 'array',
  Object = 'object',
}

export enum SchemaCombinerName {
  AllOf = 'allOf',
  AnyOf = 'anyOf',
  OneOf = 'oneOf',
}

export type SchemaAnnotations = 'title' | 'description' | 'default' | 'examples';

export type JSONSchema4Metadata = 'id' | '$schema';
