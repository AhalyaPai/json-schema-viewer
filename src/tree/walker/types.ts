import { Dictionary, JsonPath } from '@stoplight/types';
import { JSONSchema4 } from 'json-schema';
import { SchemaNode } from './nodes/types';

export type SchemaFragment = Dictionary<unknown, keyof JSONSchema4>;

export type WalkerRefResolver = (path: JsonPath | null, $ref: string) => JSONSchema4;

export type WalkingOptions = {
  mergeAllOf: boolean;
  onNode?(node: SchemaNode, level: number): boolean | void;
  stepIn?: boolean;
  resolveRef: WalkerRefResolver;
  shouldResolveEagerly: boolean;
};
