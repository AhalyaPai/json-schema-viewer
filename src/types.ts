export type ViewMode = 'read' | 'write' | 'standalone';
