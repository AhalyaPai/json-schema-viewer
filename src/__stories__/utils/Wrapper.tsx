import * as React from 'react';

export const Wrapper: React.FunctionComponent<React.HTMLAttributes<HTMLElement>> = ({ children, ...props }) => {
  return (
    <div style={{ margin: '200px auto', width: '80vw' }}>
      <div
        // @ts-ignore
        style={{
          width: '100%',
          position: 'relative',
          height: '600px',
          padding: 20,
        }}
        {...props}
      >
        {children}
      </div>
    </div>
  );
};
