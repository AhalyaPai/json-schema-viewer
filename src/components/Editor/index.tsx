import { isRootNode, SchemaNodeKind } from '@stoplight/json-schema-tree';
import { Button, Input, Select } from '@stoplight/mosaic';
import { Checkbox } from '@stoplight/ui-kit';
import { capitalize } from 'lodash';
import { action } from 'mobx';
import { observer } from 'mobx-react-lite';
import * as React from 'react';

import { useActiveNodeContext } from '../../contexts';

const OPTIONS = Object.values(SchemaNodeKind).map(kind => ({ value: kind }));

const FormRow: React.FC<{ className?: string; label: string; inline?: boolean }> = ({
  className,
  label,
  inline = true,
  children,
}) => {
  return (
    <div
      className={`sl-flex ${inline ? 'sl-items-center' : 'sl-items-start sl-flex-col'} sl-w-full sl-my-2 ${
        className ?? ''
      }`}
    >
      <label className="sl-font-bold sl-text-md sl-pr-2">{capitalize(label)}:</label>
      {children}
    </div>
  );
};

export const Editor = observer(() => {
  const node = useActiveNodeContext();

  if (!node) {
    return <div className="sl-px-4 sl-w-80">Please select a node</div>;
  }

  const types = node.types;
  const prop = node.subpath[node.subpath.length - 1];
  const children = node?.parent?.children;
  const index = node?.parent?.children?.indexOf(node);

  return (
    <div className="sl-px-4 sl-w-80">
      ID: {node.id.replace('json-schema-tree-id-', '')}
      <FormRow label="name" inline>
        <Input
          className="sl-w-full"
          placeholder="name"
          value={prop}
          onChange={action(e => {
            node.subpath[node.subpath.length - 1] = e.target.value;
          })}
        />
      </FormRow>
      <FormRow label="title">
        <Input
          className="sl-w-full"
          placeholder="title"
          value={String(node.title ?? '')}
          onChange={action(e => {
            node.title = e.target.value;
          })}
        />
      </FormRow>
      <FormRow label="description" inline={false}>
        <textarea
          className="sl-w-full"
          placeholder="description"
          style={{ appearance: 'none' }}
          value={String(node.annotations.description ?? '')}
          onChange={action(e => {
            node.annotations.description = e.target.value;
          })}
        />
      </FormRow>
      <FormRow label="type">
        <Select
          value={types.values().next().value}
          options={OPTIONS}
          onChange={action(e => {
            types.clear();
            types.add(e);
          })}
        />
      </FormRow>
      <FormRow label="deprecated">
        <Checkbox
          checked={!!node.annotations.deprecated}
          onChange={action(e => {
            node.annotations.deprecated = e.target.checked;
          })}
        />
      </FormRow>
      <FormRow label="required">
        <Checkbox
          checked={!!node.parent?.validations?.required?.includes(prop)}
          onChange={action(e => {
            e.target.checked
              ? node.parent?.validations?.required.push(prop)
              : node.parent?.validations?.required.delete(prop);
          })}
        />
      </FormRow>
      {children ? (
        <div>
          <Button
            onPress={action(() => {
              children[index] = children.splice(index - 1, 1, children[index])[0];
            })}
            disabled={index === 0}
          >
            Move up
          </Button>
          <Button
            onPress={action(() => {
              children[index] = children.splice(index + 1, 1, children[index])[0];
            })}
            disabled={index + 1 === children.length}
          >
            Move down
          </Button>
          <Button>Move under</Button>
        </div>
      ) : null}
      <Button
        onPress={() => {
          console.log(JSON.stringify(node, null, 2));
        }}
      >
        Export node
      </Button>
      <Button
        onPress={() => {
          let a = node;
          while (!isRootNode(a)) {
            a = a?.parent;
          }
          console.log(JSON.stringify(a, null, 2));
        }}
      >
        Export model
      </Button>
    </div>
  );
});
