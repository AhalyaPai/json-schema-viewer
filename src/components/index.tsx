export * from './JsonSchemaViewer';
export * from './SchemaRow';
export * from './SchemaTree';
export * from './Type';
export * from './Types';
