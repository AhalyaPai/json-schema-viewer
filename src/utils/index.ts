export * from './isPropertyRequired';
export * from './printName';
export * from './isParentNode';
