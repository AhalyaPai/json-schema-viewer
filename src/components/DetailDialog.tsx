import * as cn from 'classnames';
import _get = require('lodash/get');
import _isEmpty = require('lodash/isEmpty');
import { observer } from 'mobx-react-lite';
import * as React from 'react';

import { MarkdownViewer } from '@stoplight/markdown-viewer';
import { TreeStore } from '@stoplight/tree-list';
import { Dialog } from '@stoplight/ui-kit';

import { SchemaNodeWithMeta } from '../types';
import { isCombiner, isRef } from '../utils';
import { Types } from './';

export interface IDetailDialog extends React.HTMLAttributes<HTMLDivElement> {
  treeStore: TreeStore;
}

export const DetailDialog = observer<IDetailDialog>(({ treeStore }) => {
  const onClose = React.useCallback(() => treeStore.setActiveNode(''), []);

  if (!treeStore.activeNodeId) return null;

  const node = treeStore.nodes.find(n => n.id === treeStore.activeNodeId);
  if (!node) return null;

  const meta = node.metadata as SchemaNodeWithMeta;
  const { name, subtype, $ref, required } = meta;

  const type = isRef(meta) ? '$ref' : isCombiner(meta) ? meta.combiner : meta.type;
  const description = _get(meta, 'annotations.description', 'No further description.');

  const validations = 'validations' in meta && meta.validations ? meta.validations : [];
  const validationElems = [];
  for (const key in validations) {
    if (!validations[key]) continue;

    validationElems.push(
      <div key={key} className="flex py-1">
        <div className="flex-1">{key}:</div>
        <div className="pl-10">{validations[key] as any}</div>
      </div>
    );
  }

  return (
    <Dialog
      isOpen
      onClose={onClose}
      title={
        <div className="py-3">
          <div className="flex items-center text-base">
            {name && <span className="mr-3 te">{name}</span>}

            <Types type={type} subtype={subtype}>
              {type === '$ref' ? `[${$ref}]` : null}
            </Types>
          </div>

          <div className={cn('text-xs font-semibold', required ? 'text-red-6' : 'text-darken-7')}>
            {required ? 'REQUIRED' : 'OPTIONAL'}
          </div>
        </div>
      }
    >
      <div className="px-6 text-sm flex">
        {description && (
          <div className="flex-1">
            <MarkdownViewer className="mt-6" markdown={description} />
          </div>
        )}

        {!_isEmpty(validationElems) && <div className="mt-4 pl-4 border-l py-2">{validationElems}</div>}
      </div>
    </Dialog>
  );
});
DetailDialog.displayName = 'JsonSchemaViewer.DetailDialog';
