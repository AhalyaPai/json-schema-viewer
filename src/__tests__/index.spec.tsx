import 'jest-enzyme';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as fastGlob from 'fast-glob';
import * as fs from 'fs';
import { JSONSchema4 } from 'json-schema';
import { curry } from 'lodash';
import * as path from 'path';
import * as React from 'react';

import { JsonSchemaViewer } from '../components';
import { ViewMode } from '../types';
import { dumpDom } from './utils/dumpDom';

describe('HTML Output', () => {
  it.each(
    fastGlob.sync('**/*.json', {
      cwd: path.join(__dirname, '../__fixtures__'),
      ignore: ['stress-schema.json'],
    }),
  )('should match %s', filename => {
    const schema = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../__fixtures__/', filename), 'utf8'));

    expect(dumpDom(<JsonSchemaViewer schema={schema} defaultExpandedDepth={Infinity} />)).toMatchSnapshot();
  });

  describe.each(['anyOf', 'oneOf'])('given %s combiner placed next to allOf', combiner => {
    let schema: JSONSchema4;

    beforeEach(() => {
      schema = {
        type: 'object',
        title: 'Account',
        allOf: [
          {
            type: 'object',
            properties: {
              type: {
                type: 'string',
                enum: ['admin', 'editor'],
              },
              enabled: {
                type: 'boolean',
                description: 'Is this account enabled',
              },
            },
            required: ['type'],
          },
        ],
        [combiner]: [
          {
            type: 'object',
            title: 'Admin',
            properties: {
              root: {
                type: 'boolean',
              },
              group: {
                type: 'string',
              },
              expirationDate: {
                type: 'string',
              },
            },
          },
          {
            type: 'object',
            title: 'Editor',
            properties: {
              supervisor: {
                type: 'string',
              },
              key: {
                type: 'string',
              },
            },
          },
        ],
      };
    });

    it('given allOf merging enabled, should merge contents of allOf combiners', () => {
      expect(dumpDom(<JsonSchemaViewer schema={schema} defaultExpandedDepth={Infinity} />)).toMatchSnapshot();
    });
  });

  it('given array with oneOf containing items, should merge it correctly', () => {
    const schema: JSONSchema4 = {
      oneOf: [
        {
          items: {
            type: 'string',
          },
        },
        {
          items: {
            type: 'number',
          },
        },
      ],
      type: 'array',
    };

    render(<JsonSchemaViewer schema={schema} defaultExpandedDepth={Infinity} />);

    expect(screen.queryByText('array[string]')).toBeInTheDocument();
    expect(screen.queryByText('array[number]')).toBeInTheDocument();
  });

  it.each<ViewMode>(['standalone', 'read', 'write'])('given %s mode, should populate proper nodes', mode => {
    const schema: JSONSchema4 = {
      type: ['string', 'object'],
      properties: {
        id: {
          type: 'string',
          readOnly: true,
        },
        description: {
          type: 'string',
          writeOnly: true,
        },
      },
    };

    render(<JsonSchemaViewer schema={schema} defaultExpandedDepth={Infinity} viewMode={mode} />);

    const idElement = screen.queryByText('id');
    const descriptionElement = screen.queryByText('description');
    if (mode !== 'read') {
      expect(descriptionElement).toBeInTheDocument();
    }
    if (mode !== 'write') {
      expect(idElement).toBeInTheDocument();
    }
  });

  it('given multiple object and string type, should process properties', () => {
    const schema: JSONSchema4 = {
      type: ['string', 'object'],
      properties: {
        ids: {
          type: 'array',
          items: {
            type: 'integer',
          },
        },
      },
    };

    render(<JsonSchemaViewer schema={schema} defaultExpandedDepth={Infinity} />);

    expect(screen.queryByText('ids')).toBeInTheDocument();
  });

  it('given complex type that includes array and complex array subtype, should not ignore subtype', () => {
    const description =
      "This description can be long and should truncate once it reaches the end of the row. If it's not truncating then theres and issue that needs to be fixed. Help!";
    const schema: JSONSchema4 = {
      type: 'object',
      properties: {
        items: {
          type: ['null', 'array'],
          items: {
            type: ['string', 'number'],
          },
          description: description,
        },
      },
    };

    render(<JsonSchemaViewer schema={schema} defaultExpandedDepth={Infinity} />);

    expect(screen.queryByText('array[string,number]')).toBeInTheDocument();
    expect(screen.queryByText(description)).toBeInTheDocument();
  });

  it('given visible $ref node, should try to inject the title immediately', () => {
    const schema: JSONSchema4 = {
      type: 'object',
      properties: {
        foo: {
          $ref: '#/properties/user',
        },
        user: {
          title: 'User',
          type: 'object',
          properties: {
            name: {
              type: 'string',
            },
          },
        },
      },
    };

    render(<JsonSchemaViewer schema={schema} />);

    expect(screen.queryAllByText('User')).toHaveLength(2); // one for user, one for foo
  });
});

describe.each([{}, { unknown: '' }, { $ref: null }])('given empty schema, should render empty text', schema => {
  const { container } = render(<JsonSchemaViewer schema={schema as any} />);
  expect(container).toHaveTextContent('No schema defined');
});

describe('Expanded depth', () => {
  /**
   * Makes sure that exactly the first `desiredVisibleElementCount` elements from `hierarchy` are visible on the screen.
   */
  const assertTreeFragmentVisible = curry((hierarchy: readonly string[], desiredVisibleElementCount: number) => {
    for (const hierarchyElement of hierarchy.slice(0, desiredVisibleElementCount)) {
      expect(screen.getByText(hierarchyElement)).toBeInTheDocument();
    }
    if (hierarchy.length > desiredVisibleElementCount) {
      expect(screen.queryByText(hierarchy[desiredVisibleElementCount])).not.toBeInTheDocument();
    }
  });

  /**
   * Expands the tree, starting with the `expandRangeStart` element (first one clicked)
   * and stopping at the `expandRangeEnd` (not clicked).
   *
   * Finally ensures that the desired portion of the tree is visible.
   */
  const expandAndAssertTreeFragmentVisible = curry(
    (hierarchy: readonly string[], expandRangeStart: number, expandRangeEnd: number) => {
      for (const hierarchyElement of hierarchy.slice(expandRangeStart, expandRangeEnd)) {
        const element = screen.getByText(hierarchyElement);

        userEvent.click(element!);
      }

      assertTreeFragmentVisible(hierarchy, expandRangeEnd + 1);
    },
  );

  const mergedArrayWithObjectSchema = {
    type: 'array',
    items: {
      type: 'object',
      title: 'outer',
      properties: {
        foo: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              bar: {
                type: 'object',
                properties: {
                  baz: {
                    type: 'integer',
                  },
                },
              },
            },
          },
        },
      },
    },
  };

  const mergedArrayWithObjectNo2Schema = {
    type: 'array',
    items: {
      type: 'object',
      title: 'outer',
      properties: {
        bar: {
          type: 'integer',
        },
        foo: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              bar2: {
                type: 'string',
              },
              foo2: {
                type: 'number',
              },
            },
          },
        },
      },
    },
  };

  const nestedObjectSchema = {
    type: 'object',
    title: 'root',
    properties: {
      bar: {
        type: 'object',
        properties: {
          barFoo: {
            type: 'object',
            properties: {
              barFooBar: {
                type: 'object',
              },
            },
          },
          barBar: {
            type: 'string',
          },
          barBaz: {
            type: 'boolean',
          },
        },
      },
      foo: {
        type: 'array',
        items: {
          type: 'object',
          properties: {
            fooBar: {
              type: 'string',
            },
            fooFoo: {
              type: 'number',
            },
          },
        },
      },
    },
  };

  describe.each([
    ['merged array with object', mergedArrayWithObjectSchema, ['outer[]', 'foo', 'bar', 'baz']],
    ['merged array with object #2', mergedArrayWithObjectNo2Schema, ['outer[]', 'foo', 'bar2']],
    ['nested object', nestedObjectSchema, ['root', 'bar', 'barFoo', 'barFooBar']],
  ])('%s', (title, schema, hierarchy) => {
    const assertVisibleCount = assertTreeFragmentVisible(hierarchy);
    const tryExpandAndAssert = expandAndAssertTreeFragmentVisible(hierarchy);

    describe('static', () => {
      it.each([-1, 0, 1])('defaultExpandedDepth={%i}', level => {
        render(<JsonSchemaViewer schema={schema as JSONSchema4} defaultExpandedDepth={level} />);

        assertVisibleCount(level + 2);
      });
    });

    describe('actual expanding', () => {
      it.each([-1, 0, 1])('starting from defaultExpandedDepth={%i}, should expand successfully', level => {
        render(<JsonSchemaViewer schema={schema as JSONSchema4} defaultExpandedDepth={level} />);

        tryExpandAndAssert(level + 1, hierarchy.length - 1);
      });
    });
  });
});

describe('$ref resolving', () => {
  it('should render type for schema with top-level $ref pointing at primitive type', () => {
    const schema: JSONSchema4 = {
      $ref: '#/definitions/foo',
      definitions: {
        foo: {
          type: 'string',
        },
      },
    };

    render(<JsonSchemaViewer schema={schema} />);

    expect(screen.getByText('string')).toBeInTheDocument();
  });

  it('should render caret for top-level array with $ref items', () => {
    const schema: JSONSchema4 = {
      type: 'array',
      items: {
        $ref: '#/foo',
      },
    };

    render(<JsonSchemaViewer schema={schema} />);

    expect(screen.getByText('$ref(#/foo)[]')).toBeInTheDocument();
  });
});
