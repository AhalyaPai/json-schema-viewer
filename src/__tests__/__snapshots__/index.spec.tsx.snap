// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML Output should match arrays/of-allofs.json 1`] = `"Test{1}array-all-objectsarray[object]{2}foostringbarstring"`;

exports[`HTML Output should match arrays/of-arrays.json 1`] = `"array[object]{2}barintegerfooarray[array]{0}"`;

exports[`HTML Output should match arrays/of-objects.json 1`] = `"object{1}propertyIsArrayOfObjectsarray[object]{1}ArrayObjectPropertystring"`;

exports[`HTML Output should match arrays/of-refs.json 1`] = `"$ref(./models/todo-full.json)[]Cannot dereference external references"`;

exports[`HTML Output should match arrays/with-multiple-arrayish-items.json 1`] = `"array{2}numberobject{3}codenumberrequiredmsgstringrequiredrefstring"`;

exports[`HTML Output should match arrays/with-ordered-items.json 1`] = `"array{2}numberstring"`;

exports[`HTML Output should match arrays/with-single-arrayish-items.json 1`] = `"array[object]{3}codenumberrequiredmsgstringrequiredrefstring"`;

exports[`HTML Output should match combiners/allOfs/base.json 1`] = `"object{5}AllOfMergeObjectsobject{2}Object1Propertystring>= 1 charactersObject2Propertynumber<= 2AllOfMergeValidationsstring>= 1 characters<= 2 charactersAllOfMergeTakeMoreLogicalValidationnumber<= 1AllOfMergeObjectPropertyValidationsobject{1}Propertystring>= 1 characters<= 2 charactersAllOfMergeRefsobject{4}street_addressstringrequiredcitystringrequiredstatestringrequiredzipCodestring"`;

exports[`HTML Output should match combiners/allOfs/complex.json 1`] = `"object{2}fooobject{1}userallOfobject{2}names$ref(#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/name)[]Could not resolve '#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/name'usersarray[object]{3}creationobject{1}userallOfobject{2}names$ref(#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/name)[]Could not resolve '#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/name'usersarray[object]{3}creationobject{1}userallOfobject{2}names$ref(#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/name)[]Could not resolve '#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/name'usersarray{1}foo#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contactsCould not resolve '#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contacts'products#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contactsCould not resolve '#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contacts'foo#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contactsCould not resolve '#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contacts'products#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contactsCould not resolve '#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contacts'"`;

exports[`HTML Output should match combiners/allOfs/todo-full.json 1`] = `"Todo Partial[]{7}namestringrequiredcompletedboolean or nullrequiredidintegerrequired>= 0<= 1000000completed_atstring or null<date-time>created_atstring<date-time>updated_atstring<date-time>userUser{2}requirednamestringrequiredThe user's full name.agenumberrequired>= 0<= 150"`;

exports[`HTML Output should match combiners/allOfs/todo-full-2.json 1`] = `"Todo Full{5}teststringidintegerrequired>= 0<= 1000000completed_atstring or null<date-time>created_atstring<date-time>updated_atstring<date-time>"`;

exports[`HTML Output should match combiners/allOfs/with-type.json 1`] = `"oneOfBugExample{6}An example model to demonstrate a bug.actionTypestringrequiredAllowed values:\\"Cancel\\"\\"Confirm\\"\\"Update\\"idstringrequiredThe identifier of the existing reservation.externalIdstringcalculateCostsbooleancalculateDatesbooleanitemsarray[string]requiredorBugExample{6}An example model to demonstrate a bug.actionTypestringrequiredAllowed value:\\"Create\\"idstringThe identifier of the existing reservation.externalIdstringcalculateCostsbooleancalculateDatesbooleanitemsarray[string]required"`;

exports[`HTML Output should match combiners/oneof-with-array-type.json 1`] = `"oneOfarray[object]{2}foostringAllowed value:\\"test\\"bazintegerorarray[object]{3}foonumberbazintegerbarstring"`;

exports[`HTML Output should match default-schema.json 1`] = `"User{11}namestringrequiredThe user's full name. This description can be long and should truncate once it reaches the end of the row. If it's not truncating then theres and issue that needs to be fixed. Help!Example values:\\"Example name\\"\\"Different name\\"Allowed value:\\"Constant name\\"agenumberread-onlyrequired>= 10<= 40Allowed values:10203040Default value:20Multiple of:10completed_atstring<date-time>write-onlyrequiredMatch pattern:^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$itemsnull or array[string,number]This description can be long and should truncate once it reaches the end of the row. If it's not truncating then theres and issue that needs to be fixed. Help!>= 1 items<= 4 itemsemailstring<email>deprecated>= 2 charactersDefault value:\\"default@email.com\\"Example values:\\"one@email.com\\"\\"two@email.com\\"plananyOfobject{2}Example value:\\"hi\\"foostringrequiredbarstringrequired"`;

exports[`HTML Output should match formats-schema.json 1`] = `"model-with-formats{5}date-of-birthnumber or string or array<date-time>namestringidnumber<float>notype<date-time>permissionsstring or object{1}<password>idsarray[integer]"`;

exports[`HTML Output should match references/base.json 1`] = `"object{2}billing_addressobject{3}street_addressstringrequiredcitystringrequiredstatestringrequiredshipping_addressobject{3}street_addressstringrequiredcitystringrequiredstatestringrequired"`;

exports[`HTML Output should match references/nullish.json 1`] = `"objectempty-ref$ref$ref is not a string"`;

exports[`HTML Output should match tickets.schema.json 1`] = `"object{3}This section allows the selection of the ticketing options for all sales of the order.availableTicketingOptions$ref(../TicketingOptionInfo/TicketingOptionInfo.v1-0.yaml)[]List of ticketing options of the order.Cannot dereference external referencescommonTicketingOptionsarray[string]Common ticketing options to all order items.ticketingOptionChoicearray[object]{2}Ticketing option selection per order item.statestringThe status that addresses if a specific ticket option is active or not. The status active is used before ticketing or before exchange confirmation. After ticketing, the status changes in completed. This allows to store ticketing options already used at ticketing time and to clean up all non selected options after ticketing or exchanged confirmation.Allowed values:\\"COMPLETED\\"\\"ACTIVE\\"Default value:\\"ACTIVE\\"orderItemBreakdownarray[object]{2}Structure that contains ticketing options per order item.orderItemIdstring<uuid>optionsarray[object]{4}Available ticketing options for a given order item.titlestringread-onlyTicketing option short-description.Allowed values:\\"HOMEPRINT\\"\\"TICKETLESS\\"\\"PRINT_AT_KIOSK\\"\\"SECURE_PAPER\\"selectedbooleanFlag to specify which ticketing option is selected. Only one option is allowed to be selected.exampleadditionalRequiredInfostringAdditional passenger required info specific to the given ticketing option.deliveryInfoobject{5}Data for ticket delivery."`;
